### PR TITLE
Add "cookie" property to rpc http messages

### PR DESCRIPTION
**Pull request by mhoeger**
Protobuf PR: https://github.com/Azure/azure-functions-language-worker-protobuf/pull/32
Node worker PR: https://github.com/Azure/azure-functions-nodejs-worker/pull/208

Resolves https://github.com/Azure/azure-functions-host/issues/2737

### CODE REVIEW

File: src/WebJobs.Script.Grpc/azure-functions-language-worker-protobuf/src/proto/shared/NullableTypes.proto [hunk 1]
```diff
@@ -0,0 +1,28 @@
...
+
+import "google/protobuf/timestamp.proto";
+
+message NullableString {
+	oneof string {
+		string value = 1;
+	}
+}
+
+message NullableDouble {
```
Comment (pragnagopa) [inline src/WebJobs.Script.Grpc/azure-functions-language-worker-protobuf/src/proto/shared/NullableTypes.proto]: Please do a protobuf release and pull from the release tag.

File: src/WebJobs.Script/Binding/Http/HttpBinding.cs [hunk 2]
```diff
@@ -14,6 +14,8 @@
 using Microsoft.AspNetCore.Mvc;
 using Microsoft.AspNetCore.Mvc.WebApiCompatShim;
 using Microsoft.Azure.WebJobs.Script.Description;
+using Microsoft.Azure.WebJobs.Script.Grpc.Messages;
```
Comment (pragnagopa) [inline src/WebJobs.Script/Binding/Http/HttpBinding.cs]: Do you need reference to Microsoft.Azure.WebJobs.Script.Grpc.Messages? It is best to confine reference to Microsoft.Azure.WebJobs.Script.Grpc.Messages to Rpc layer to make it easier to switch Rpc implementation if needed,

File: src/WebJobs.Script/Rpc/MessageExtensions/RpcMessageConversionExtensions.cs [hunk 3]
```diff
@@ -88,7 +88,10 @@ public static TypedData ToRpc(this object value, ILogger logger)
 
                 foreach (var pair in request.Headers)
                 {
-                    http.Headers.Add(pair.Key.ToLowerInvariant(), pair.Value.ToString());
+                    if (!string.IsNullOrEmpty(pair.Value))
+                    {
```
Comment (pragnagopa) [inline src/WebJobs.Script/Rpc/MessageExtensions/RpcMessageConversionExtensions.cs]: Looks like empty headers are valid: https://www.w3.org/Protocols/rfc2616/rfc2616-sec2.html#sec2.1 Do we need this change?

File: src/WebJobs.Script/Rpc/MessageExtensions/Utilities.cs [hunk 4]
```diff
@@ -24,12 +27,75 @@ public static object ConvertFromHttpMessageToExpando(RpcHttp inputMessage)
...
             if (inputMessage.Body != null)
             {
                 expando.body = inputMessage.Body.ToObject();
             }
             return expando;
         }
+
+        public static Tuple<string, string, CookieOptions> RpcHttpCookieConverter(RpcHttpCookie cookie)
+        {
+            var cookieOptions = new CookieOptions();
```
Comment (pragnagopa) [inline src/WebJobs.Script/Rpc/MessageExtensions/Utilities.cs]: Add unit test with valid values and null

File: test/WebJobs.Script.Tests/Binding/HttpBindingTests.cs [hunk 5]
```diff
@@ -75,7 +77,7 @@ public void ParseResponseObject_StatusWithNullBody_ReturnsExpectedResult()
             IDictionary<string, object> headers = null;
             int statusCode = StatusCodes.Status200OK;
             bool enableContentNegotiationResponse = false;
-            HttpBinding.ParseResponseObject(responseObject, ref content, out headers, out statusCode, out enableContentNegotiationResponse);
+            HttpBinding.ParseResponseObject(responseObject, ref content, out headers, out statusCode, out List<Tuple<string, string, CookieOptions>> cookies, out enableContentNegotiationResponse);
```
Comment (pragnagopa) [inline test/WebJobs.Script.Tests/Binding/HttpBindingTests.cs]: - Update test to verify cookies
- Add test that sets and does not set cookies

File: src/WebJobs.Script/Binding/Http/RawScriptResult.cs [hunk 6]
```diff
@@ -73,6 +76,14 @@ public async Task ExecuteResultAsync(ActionContext context)
                 response.StatusCode = StatusCode.Value;
             }
 
+            if (Cookies != null)
+            {
```
Comment (pragnagopa) [inline src/WebJobs.Script/Binding/Http/RawScriptResult.cs]: Add test to verify RawScriptResult with cookies

File: src/WebJobs.Script/Rpc/MessageExtensions/RpcMessageConversionExtensions.cs (hunk 3 above)
Comment (mhoeger) [reply src/WebJobs.Script/Rpc/MessageExtensions/RpcMessageConversionExtensions.cs]: This is in response to this issue: https://github.com/Azure/azure-functions-language-worker-protobuf/issues/21

But it is a breaking change for Java that does this correctly... I'll remove this until we solve the Buffer overrun problem for all languages.

Comment (pragnagopa) [reply src/WebJobs.Script/Rpc/MessageExtensions/RpcMessageConversionExtensions.cs]: Is it possible to handle this in nodejs worker?

Comment (mhoeger) [reply src/WebJobs.Script/Rpc/MessageExtensions/RpcMessageConversionExtensions.cs]: No, at that point the buffer overrun has already happened and the message is corrupted :( we could remove empty headers if the language is node though..??

File: src/WebJobs.Script/Binding/Http/HttpBinding.cs [hunk 7]
```diff
@@ -120,6 +122,11 @@ internal static void ParseResponseObject(IDictionary<string, object> responseObj
                 {
                     enableContentNegotiation = enableContentNegotiationValue;
                 }
+
+                if (responseObject.TryGetValue("cookies", out List<Tuple<string, string, CookieOptions>> cookiesValue, ignoreCase: true))
```
Comment (pragnagopa) [inline src/WebJobs.Script/Binding/Http/HttpBinding.cs]: I see we are using strings `cookies`, `body` etc through out this file. We should move them separate file `Httpconstants` file. It is out-of-scope for this PR. Please open a different issue.

File: test/WebJobs.Script.Tests/Binding/HttpBindingTests.cs [hunk 8]
```diff
@@ -56,12 +52,45 @@ public void ParseResponseObject_ReturnsExpectedResult()
...
+
+            object content = null;
+            var statusCode = StatusCodes.Status200OK;
+            HttpBinding.ParseResponseObject(responseObject, ref content, out IDictionary<string, object> headers, out statusCode, out List<Tuple<string, string, CookieOptions>> cookies, out bool enableContentNegotiationResponse);
+
+            Assert.Equal("Test Body", content);
+            Assert.Same(cookieContents, cookies);
+            Assert.Equal(StatusCodes.Status202Accepted, statusCode);
+            Assert.False(enableContentNegotiationResponse);
+            Assert.Same(headers, headers);
```
Comment (pragnagopa) [inline test/WebJobs.Script.Tests/Binding/HttpBindingTests.cs]: typo? headers?
Nit: change assert statements to be more explicit. For example, expectedHeaders actualHeaders

File: test/WebJobs.Script.Tests/Binding/HttpBindingTests.cs [hunk 9]
```diff
@@ -56,12 +52,45 @@ public void ParseResponseObject_ReturnsExpectedResult()
...
+            responseObject.Body = "Test Body";
+            responseObject.Cookies = cookieContents;
+            responseObject.StatusCode = "202";  // verify string works as well
+
+            object content = null;
+            var statusCode = StatusCodes.Status200OK;
+            HttpBinding.ParseResponseObject(responseObject, ref content, out IDictionary<string, object> headers, out statusCode, out List<Tuple<string, string, CookieOptions>> cookies, out bool enableContentNegotiationResponse);
+
+            Assert.Equal("Test Body", content);
+            Assert.Same(cookieContents, cookies);
```
Comment (pragnagopa) [inline test/WebJobs.Script.Tests/Binding/HttpBindingTests.cs]: Verify cookie options?

File: test/WebJobs.Script.Tests/Binding/HttpBindingTests.cs [hunk 10]
```diff
@@ -55,7 +54,7 @@ public void ParseResponseObject_ReturnsExpectedResult()
             HttpBinding.ParseResponseObject(responseObject, ref content, out headers, out statusCode, out cookies, out enableContentNegotiationResponse);
 
             Assert.Equal("Test Body", content);
-            Assert.Same(headers, headers);
+            Assert.Equal(null, headers);
```
Comment (pragnagopa) [inline test/WebJobs.Script.Tests/Binding/HttpBindingTests.cs]: change to Assert.Null

File: test/WebJobs.Script.Tests/Binding/HttpBindingTests.cs [hunk 11]
```diff
@@ -88,9 +87,17 @@ public void ParseResponseObject_WithCookies_ReturnsExpectedResult()
...
+            Assert.Same(cookieProperties, firstCookie);
+            Assert.Same(cookieProperties.Item1, firstCookie.Item1);
+            Assert.Same(cookieProperties.Item2, firstCookie.Item2);
+            Assert.Same(cookieProperties.Item3, firstCookie.Item3);
+            Assert.Same(cookieContents.First(), cookies.First());
+            Assert.Same(cookieContents.First(), cookies.First());
             Assert.Equal(StatusCodes.Status202Accepted, statusCode);
             Assert.False(enableContentNegotiationResponse);
-            Assert.Same(headers, headers);
+            Assert.Equal(null, headers);
```
Comment (pragnagopa) [inline test/WebJobs.Script.Tests/Binding/HttpBindingTests.cs]: change to Assert.Null

File: test/WebJobs.Script.Tests/Binding/HttpBindingTests.cs [hunk 12]
```diff
@@ -88,9 +87,17 @@ public void ParseResponseObject_WithCookies_ReturnsExpectedResult()
...
             Assert.Equal("Test Body", content);
             Assert.Same(cookieContents, cookies);
+            Assert.Equal(cookieContents.Count, cookies.Count);
+            var firstCookie = cookies.First();
+            Assert.Same(cookieProperties, firstCookie);
+            Assert.Same(cookieProperties.Item1, firstCookie.Item1);
+            Assert.Same(cookieProperties.Item2, firstCookie.Item2);
+            Assert.Same(cookieProperties.Item3, firstCookie.Item3);
+            Assert.Same(cookieContents.First(), cookies.First());
+            Assert.Same(cookieContents.First(), cookies.First());
```
Comment (pragnagopa) [inline test/WebJobs.Script.Tests/Binding/HttpBindingTests.cs]: Dup asserts?
